### PR TITLE
Remove InitializeNonlocalInstruction

### PR DESCRIPTION
**Pull request by rdmarsh2**
This removes `InitializeNonlocalInstruction` from the IR, which should result in a small performance improvement in building the IR. I think the consistency checks need to be adjusted to handle phi nodes for global and static variables which don't have their address taken, but I'm not sure exactly how to identify those.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll [hunk 1]
```diff
@@ -81,8 +81,11 @@ private newtype TMemoryLocation =
   TAllNonLocalMemory(IRFunction irFunc, boolean isMayAccess) {
     isMayAccess = false or isMayAccess = true
   } or
-  TAllAliasedMemory(IRFunction irFunc, boolean isMayAccess) {
-    isMayAccess = false or isMayAccess = true
+  TAllAliasedMemory(IRFunction irFunc, boolean isMayAccess, boolean canDefineReadOnly) {
+    isMayAccess = false and
+    (canDefineReadOnly = true or canDefineReadOnly = false)
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll:86]:
```suggestion
    canDefineReadOnly = [true, false]
```
I think this is easier to read, making it clear that it's the same variable in both cases. That's in contrast to the line below, where it's different variables bound to true and false.

**Review by jbj — COMMENTED** [review]
It's great to see that this diff takes away more than it adds.

**Comment by MathiasVP** [comment]
Yay for fewer instructions 🎉 ! Apparently, GitHub doesn't allow comments on collapsed lines, so I'm leaving this as a comment instead:

I guess now that we no longer have an `InitializeNonLocal` instruction this comment on the `canDefineReadOnly` override in `AllNonLocalMemory` doesn't make sense anymore, right?
```codeql
  override predicate canDefineReadOnly() {
    // A "must" access that defines all non-local memory appears only on the `InitializeNonLocal`
    // instruction, which provides the initial definition for all memory outside of the current
    // function's stack frame. This memory includes string literals and other read-only globals, so
    // we allow such an access to be the definition for a use of a read-only location.
    not isMayAccess()
  }
```

**Comment by rdmarsh2** [comment]
Yeah, I think I can just make that `override predicate canDefineReadOnly() { none() }` now.

**Comment by jbj** [comment]
The tests are passing. Is it time to take this out of draft mode?

**Review by dbartol — DISMISSED** [review]
LGTM (@rdmarsh2 and I had already discussed most of these changes via Zoom).

**Review by dbartol — APPROVED** [review]
LGTM

**Pull request merged**